File: src/test/java/io/lsdconsulting/lsd/distributed/interceptor/captor/http/ResponseCaptorShould.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.http;

import feign.Request;
import feign.RequestTemplate;
import feign.Response;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.access.repository.InterceptedDocumentRepository;
import io.lsdconsulting.lsd.distributed.interceptor.captor.header.HeaderRetriever;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.HttpStatusDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.PathDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpResponse;

import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import static feign.Request.HttpMethod.GET;
import static java.nio.charset.Charset.defaultCharset;
import static org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.emptyString;
import static org.hamcrest.Matchers.is;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.springframework.http.HttpStatus.NO_CONTENT;

class ResponseCaptorShould {

    private final InterceptedDocumentRepository interceptedDocumentRepository = mock(InterceptedDocumentRepository.class);
    private final SourceTargetDeriver sourceTargetDeriver = mock(SourceTargetDeriver.class);
    private final TraceIdRetriever traceIdRetriever = mock(TraceIdRetriever.class);
    private final HttpRequest httpRequest = mock(HttpRequest.class);
    private final ClientHttpResponse clientHttpResponse = mock(ClientHttpResponse.class);
    private final HeaderRetriever headerRetriever = mock(HeaderRetriever.class);
    private final HttpStatusDeriver httpStatusDeriver = mock(HttpStatusDeriver.class);

    private final PathDeriver pathDeriver = new PathDeriver();

    private final ResponseCaptor underTest = new ResponseCaptor(interceptedDocumentRepository,
            sourceTargetDeriver, pathDeriver, traceIdRetriever, headerRetriever, httpStatusDeriver, "profile");

    private final String url = randomAlphanumeric(20);
    private final String body = randomAlphanumeric(20);
    private final String traceId = randomAlphanumeric(20);
    private final String target = randomAlphanumeric(20);
    private final String path = randomAlphanumeric(20);

    private final Map<String, Collection<String>> requestHeaders = Map.of("b3", List.of(traceId), "Target-Name", List.of(target));

    private final Response response = Response.builder()
            .request(Request.create(GET, url, requestHeaders, body.getBytes(), defaultCharset(), new RequestTemplate()))
            .build();

    @Test
    public void takeTraceIdFromRequestHeaders() {
        given(traceIdRetriever.getTraceId(eq(requestHeaders))).willReturn(traceId);
        given(headerRetriever.retrieve(any(Request.class))).willReturn(requestHeaders);

        final InterceptedInteraction interceptedInteraction = underTest.captureResponseInteraction(response, 10L);

        assertThat(interceptedInteraction.getTraceId(), is(traceId));
    }

    @Test
    public void deriveTargetFromRequestHeaders() {
        given(sourceTargetDeriver.deriveTarget(eq(requestHeaders), eq(url))).willReturn(target);
        given(headerRetriever.retrieve(any(Request.class))).willReturn(requestHeaders);

        final InterceptedInteraction interceptedInteraction = underTest.captureResponseInteraction(response, 10L);

        assertThat(interceptedInteraction.getTarget(), is(target));
    }

    @Test
    public void handleEmptyResponseBodyFromDeleteRequest() throws IOException {
        HttpHeaders httpHeaders = mock(HttpHeaders.class);
        given(httpRequest.getHeaders()).willReturn(httpHeaders);
        given(clientHttpResponse.getHeaders()).willReturn(httpHeaders);
        given(clientHttpResponse.getStatusCode()).willReturn(NO_CONTENT);

        final InterceptedInteraction interceptedInteraction = underTest.captureResponseInteraction(httpRequest, clientHttpResponse, target, path, traceId, 10L);

        assertThat(interceptedInteraction.getBody(), is(emptyString()));
    }
}

File: src/test/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/http/ResponseCaptorShould.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.http

import feign.Request
import feign.Request.HttpMethod.GET
import feign.RequestTemplate
import feign.Response
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.PathDeriver
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever
import io.lsdconsulting.lsd.distributed.interceptor.persistance.RepositoryService
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`
import org.junit.jupiter.api.Test
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpRequest
import org.springframework.http.HttpStatus
import org.springframework.http.client.ClientHttpResponse
import java.io.IOException
import java.nio.charset.Charset

internal class ResponseCaptorShould {
    private val repositoryService = mockk<RepositoryService>(relaxed = true)
    private val sourceTargetDeriver = mockk<SourceTargetDeriver>()
    private val traceIdRetriever = mockk<TraceIdRetriever>()
    private val httpRequest = mockk<HttpRequest>()
    private val clientHttpResponse = mockk<ClientHttpResponse>()
    private val httpHeaderRetriever = mockk<HttpHeaderRetriever>()
    private val pathDeriver = PathDeriver()

    private val underTest = ResponseCaptor(repositoryService, sourceTargetDeriver, pathDeriver, traceIdRetriever, httpHeaderRetriever, "profile")

    private val resource = randomAlphanumeric(20)
    private val url = "http://localhost/$resource"
    private val body = randomAlphanumeric(20)
    private val traceId = randomAlphanumeric(20)
    private val target = randomAlphanumeric(20)
    private val serviceName = randomAlphanumeric(20)
    private val path = randomAlphanumeric(20)
    private val requestHeaders = mapOf<String, Collection<String>>("b3" to listOf(traceId), "Target-Name" to listOf(target))
    private val responseHeaders = mapOf<String, Collection<String>>()
    private val response = Response.builder().request(Request.create(GET, url, requestHeaders, body.toByteArray(), Charset.defaultCharset(), RequestTemplate())).build()

    @Test
    fun takeTraceIdFromRequestHeaders() {
        every { traceIdRetriever.getTraceId(eq(requestHeaders)) } returns traceId
        every { sourceTargetDeriver.deriveTarget(eq(requestHeaders), eq("/$resource")) } returns target
        every { sourceTargetDeriver.deriveServiceName(requestHeaders) } returns serviceName
        every { httpHeaderRetriever.retrieve(any<Request>()) } returns requestHeaders
        every { httpHeaderRetriever.retrieve(any<Response>()) } returns responseHeaders

        val (traceId) = underTest.captureResponseInteraction(response, 10L)

        assertThat(traceId, `is`(this.traceId))
    }

    @Test
    fun deriveTargetFromRequestHeaders() {
        every { traceIdRetriever.getTraceId(eq(requestHeaders)) } returns traceId
        every { sourceTargetDeriver.deriveTarget(eq(requestHeaders), eq("/$resource")) } returns target
        every { sourceTargetDeriver.deriveServiceName(eq(requestHeaders)) } returns serviceName
        every { httpHeaderRetriever.retrieve(any<Request>()) } returns requestHeaders
        every { httpHeaderRetriever.retrieve(any<Response>()) } returns responseHeaders

        val (_, _, _, _, _, target1) = underTest.captureResponseInteraction(response, 10L)

        assertThat(target1, `is`(target))
    }

    @Test
    fun enqueueInterceptedInteractionOnFeignResponse() {
        every { traceIdRetriever.getTraceId(eq(requestHeaders)) } returns traceId
        every { sourceTargetDeriver.deriveTarget(eq(requestHeaders), eq("/$resource")) } returns target
        every { sourceTargetDeriver.deriveServiceName(eq(requestHeaders)) } returns serviceName
        every { httpHeaderRetriever.retrieve(any<Request>()) } returns requestHeaders
        every { httpHeaderRetriever.retrieve(any<Response>()) } returns responseHeaders

        val interceptedInteraction = underTest.captureResponseInteraction(response, 10L)

        verify { repositoryService.enqueue(interceptedInteraction) }
    }

    @Test
    @Throws(IOException::class)
    fun handleEmptyResponseBodyFromDeleteRequest() {
        val httpHeaders = mockk<HttpHeaders>()
        every { httpHeaders.contentLength } returns 10
        every { httpRequest.headers } returns httpHeaders
        every { clientHttpResponse.headers } returns httpHeaders
        every { clientHttpResponse.statusCode } returns HttpStatus.NO_CONTENT
        every { clientHttpResponse.body } returns "body".byteInputStream()
        every { sourceTargetDeriver.deriveServiceName(requestHeaders) } returns serviceName
        every { httpHeaderRetriever.retrieve(httpRequest) } returns requestHeaders
        every { httpHeaderRetriever.retrieve(any<ClientHttpResponse>()) } returns responseHeaders

        val (_, body) = underTest.captureResponseInteraction(httpRequest, clientHttpResponse, target, path, traceId, 10L)

        assertThat(body, `is`("body"))
    }

    @Test
    @Throws(IOException::class)
    fun enqueueInterceptedInteractionOnSpringResponse() {
        val httpHeaders = mockk<HttpHeaders>()
        every { httpHeaders.contentLength } returns 10
        every { httpRequest.headers } returns httpHeaders
        every { clientHttpResponse.headers } returns httpHeaders
        every { clientHttpResponse.statusCode } returns HttpStatus.NO_CONTENT
        every { clientHttpResponse.body } returns "".byteInputStream()
        every { httpHeaderRetriever.retrieve(httpRequest) } returns requestHeaders
        every { sourceTargetDeriver.deriveServiceName(requestHeaders) } returns serviceName
        every { httpHeaderRetriever.retrieve(any<ClientHttpResponse>()) } returns responseHeaders

        val interceptedInteraction = underTest.captureResponseInteraction(httpRequest, clientHttpResponse, target, path, traceId, 10L)

        verify { repositoryService.enqueue(interceptedInteraction) }
    }
}
